config/config.go

```diff
@@ -113,8 +116,34 @@ func WithOpenTelemetryConfiguration(cfg OpenTelemetryConfiguration) Configuratio
 	})
 }
 
+// ParseYAML parses a YAML configuration file into an OpenTelemetryConfiguration.
+func ParseYAML(file []byte) (*OpenTelemetryConfiguration, error) {
+	re := regexp.MustCompile(`\$\{([a-zA-Z_][a-zA-Z0-9_]*)\}`)
+
+	replaceEnvVars := func(input []byte) []byte {
+		return re.ReplaceAllFunc(input, func(s []byte) []byte {
+			match := re.FindSubmatch(s)
+			if len(match) < 2 {
+				return s
+			}
+			envVarName := string(match[1])
+			envVarValue := os.Getenv(envVarName)
+			return []byte(envVarValue)
+		})
+	}
+
+	file = replaceEnvVars(file)
```

Comment:
I just want to call out one important thing from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md#environment-variable-substitution:

> It MUST NOT be possible to inject YAML structures by environment variables. For example, references to INVALID_MAP_VALUE environment variable below.

Maybe you will find some code that will help with it in https://github.com/open-telemetry/opentelemetry-collector. @codeboten: are you able to provide some guidance?

Reply:
Totally agree @pellared, also this does not solve for blocking the key replacement (i.e.`${SNEAKY}="foo"`) in the spec. I will take a look at both of these cases to see if we can come up with something more useful.